statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,698 @@
+
+        if estimation_kwds is None:
+            self.estimation_kwds = {}
+        else:
+            self.estimation_kwds = esitmation_kwds
+
+        if join_method is None:
+            self.join_method = _join_debiased
+        else:
+            self.join_method = join_method
+
+        if join_kwds is None:
+            self.join_kwds = {}
+        else:
+            self.join_kwds = join_kwds
+
+
+#    def fit_partition(self, pnum, endog, exog):
+#        """handles the model fitting for each machine.
+#
+#        Parameters
+#        ----------
+#        pnum : scalar
+#            index of current partition.
+#        endog : array-like
+#            endogenous data for current partition.
+#        exog : array-like
+#            exogenous data for current partition.
+#
+#        Returns
+#        -------
+#        estimation_method result.  For the default,
+#        _est_regularized_debiased, a tuple.
+#        """
+#
+#        model = self.model_class(endog, exog, **self.init_kwds)
+#        results = self.estimation_method(model, pnum, self.partitions,
+#                                         self.fit_kwds,
+#                                         **self.estimation_kwds)
+#        return results
+
+
+    def fit_distributed(self, parallel_method="sequential"):
+        """Performs the distributed estimation using the corresponding
+        DistributedModel
+
+        Parameters
+        ----------
+        parallel_method : str
+            type of distributed estimation to be used, currently
+            "sequential", "joblib" and "dask" are supported.
+
+        Returns
+        -------
+        join_method result.  For the default, _join_debiased, it returns a
+        p length array.
+        """
+
+        if parallel_method == "sequential":
+            return self.fit_dist_sequential()
+
+        elif parallel_method == "joblib":
+            try:
+                return self.fit_dist_joblib()
+            except ImportError:
+                raise ValueError("cannot import joblib")
+
+        elif parallel_method == "dask":
+            try:
+                return self.fit_dist_dask()
+            except ImportError:
+                raise ValueError("cannot import dask")
+
+        else:
+            raise ValueError("parallel_method: %s is currently not supported"
+                             % parallel_method)
+
+
+    def fit_dist_sequential(self):
+        """Sequentially performs the distributed estimation using
+        the corresponding DistributedModel
+
+        Parameters
+        ----------
+
+        Returns
+        -------
+        join_method result.  For the default, _join_debiased, it returns a
+        p length array.
+        """
+
+        results_l = []
+
+        for pnum, (endog, exog) in enumerate(self.data_generator):
+
+            results = _helper_fit_partition(self, pnum, endog, exog)
+            results_l.append(results)
+
+        return self.join_method(results_l, self.partitions,
```

Comment:
This will be the last line of every `fit_dist_xxx` function, so better to move it into `fit_distributed` to avoid needless copying.